thoth/storages/graph/postgres.py

```diff
@@ -1645,63 +1648,141 @@ def _create_python_package_requirement(self, requirements: dict) -> List[PythonP
-
-        for package in pipfile_locked.packages.packages.values():
-            result.append(self._create_python_package_version(
-                package_name=package.name,
-                package_version=package.locked_version,
-                os_name=os_name,
-                os_version=os_version,
-                python_version=python_version,
-                index_url=package.index.url if package.index else None,
-            ))
+        if sync_only_entity:
+            result = []
+            pipfile_locked = PipfileLock.from_dict(pipfile_locked, pipfile=None)
+            os_name = software_environment.os_name if software_environment else None
+            os_version = software_environment.os_version if software_environment else None
+            python_version = software_environment.python_version if software_environment else None
+
+            for package in pipfile_locked.packages.packages.values():
+                result.append(self._create_python_package_version(
+                    package_name=package.name,
+                    package_version=package.locked_version,
+                    os_name=os_name,
+                    os_version=os_version,
+                    python_version=python_version,
+                    index_url=package.index.url if package.index else None,
+                    sync_only_entity=sync_only_entity,
+                ))
 
-        return result
+            return result
+        else:
+            result = []
+            pipfile_locked = PipfileLock.from_dict(pipfile_locked, pipfile=None)
+            os_name = software_environment.os_name
+            os_version = software_environment.os_version
+            python_version = software_environment.python_version
+
+            for package in pipfile_locked.packages.packages.values():
+                # Check if the package has a known index
+                if package.index:
+                    python_package_version = (
+                        self._session.query(PythonPackageVersion)
+                        .join(PythonPackageIndex)
+                        .filter(PythonPackageVersion.package_name == package.name)
+                        .filter(PythonPackageVersion.package_version == package.locked_version)
+                        .filter(PythonPackageVersion.os_name == os_name)
+                        .filter(PythonPackageVersion.os_version == os_version)
+                        .filter(PythonPackageVersion.python_version == python_version)
+                        .filter(PythonPackageIndex.url == package.index.url if package.index else None)
+                        .first()
+                    )
+                else:
+                    logging.warning("Package %s %s does not have Index!!!" % (package.name, package.locked_version))
+                    raise Exception("Package %s %s does not have Index!!!" % (package.name, package.locked_version))
```

Comment:
This exception is too generic please create a new one in exceptions.py file and derive from `ThothStorageException` so we can handle this specific exception in upper layers if needed.

Logging of warning before raising an exception can be removed.

And the last one - exclamation marks seem to be too imperative :), you can provide a more gentle error message (f-strings are allowed):
```python
f"Trying to sync package {package.name!r} in version {package.locked_version!r} which does not have corresponding Python entity record"
```